app/javascript/lib/scatter-plot.js

```diff
@@ -0,0 +1,308 @@
+    layout.yaxis.range = domainRanges.y
+  } else {
+    layout.xaxis.autorange = true
+    layout.yaxis.autorange = true
+  }
+
+  if (hasCoordinateLabels) {
+    layout.annotations = coordinateLabels
+  }
+
+  return layout
+}
+
+/** Set colors for group annotations */
+export function setMarkerColors(data) {
+  return data.map((trace, i) => {
+    trace.marker.color = getColorBrewerColor(i)
+    return trace
+  })
+}
+
+/** Get height and width for a to-be-rendered cluster plot */
+function calculatePlotRect(numRows, numColumns) {
+  // Get height
+  const $ideo = $('#_ideogram')
+  const ideogramHeight = $ideo.length ? $ideo.height() : 0
+  const verticalPad = 225 // Accounts for page header, search area, nav tabs
+  let height = $(window).height() - verticalPad - ideogramHeight
+  if (numRows > 1) {height = height/numRows - 10}
+
+  // Get width, and account for expanding "View Options" after page load
+  const baseWidth = $('#render-target .tab-content').actual('width')
+  const horizontalPad = 80 // Accounts for empty space to left and right
+  let width = (baseWidth - horizontalPad) / numColumns
+
+  // Ensure plots aren't too small
+  if (height < 100) {height = 100}
+  if (width < 100) {width = 100}
+
+  return { height, width }
+}
+
+/** Get layout object with various Plotly scatter plot display parameters */
+function getScatterPlotLayout(rawPlot) {
+  const { numRows, numColumns } = rawPlot
+  const { height, width } = calculatePlotRect(numRows, numColumns)
+  let layout = getBaseLayout(height, width)
+
+  let dimensionProps
+  if (rawPlot.is3D) {
+    const camera = $('#cluster-plot').data('camera')
+    dimensionProps = get3DScatterProps(camera, rawPlot)
+    layout.scene = dimensionProps
+  } else {
+    dimensionProps = get2DScatterProps(rawPlot)
+    layout = Object.assign(layout, dimensionProps)
+  }
+
+  return layout
+}
+
+/** Renders a Plotly scatter plot for "Clusters" tab */
+function renderScatterPlot(rawPlot, plotId, legendId) {
+  let { data } = rawPlot
+
+  const layout = getScatterPlotLayout(rawPlot)
+
+  if (rawPlot.annotParams.type === 'group' && !rawPlot.gene) {
+    data = setMarkerColors(data)
+  }
+
+  Plotly.newPlot(plotId, data, layout)
+
+  if (legendId) {
+    $(`#${legendId}`).html(
+      `<p class="text-center help-block">${rawPlot.description}</p>`
+    )
+  }
+
+  // access actual target div, not jQuery object wrapper for relayout event
+  const scatterPlotDiv = document.getElementById(plotId)
+  scatterPlotDiv.on('plotly_relayout', cameraData => {
+    if (typeof cameraData['scene.camera'] !== 'undefined') {
+      const newCamera = cameraData['scene.camera']
+      $(`#${plotId}`).data('camera', newCamera)
+    }
+  })
+}
+
+/**
+ * Load and draw scatter plot.  Handles clusters and spatial groups, numeric
+ * and group annotations.
+ *
+ * @param {Object} clusterParams Parameters for `/clusters` plot API endpoint
+ * @param {Object} frame Plot UI properties not from plot API data
```

Comment:
I'm not sure I understand why you're calling this param 'frame' instead of 'frameConfig'.  'frame' sounds like the value should be a DOM container.  Or maybe containerConfig?

Reply:
I've renamed these props `apiParams` and `props` to make them a bit more generically applicable.